File: bindings/swift/Sources/Causalontology/SchemaValidator.swift
// SchemaValidator.swift
//
// Schema validation against spec/schema/*.schema.json.
//
// A deliberately small interpreter for exactly the JSON Schema keywords the
// eight Causalontology schemas use: type, const, enum, pattern, required,
// properties, additionalProperties, items, minItems, minLength, minimum,
// maximum, oneOf, and local $ref (#/$defs/...). "format" is treated as an
// annotation, as the 2020-12 draft does by default.

import Foundation

/// Loads the eight schemas from a spec/schema directory and validates
/// objects against them. Not thread-safe (the schema cache is unlocked);
/// the conformance harness is single-threaded.
public final class SchemaValidator {
    /// The schema file name per kind. Three token kinds keep their original
    /// 1.0.0-reserved file names (individual/token/state); the id scheme is
    /// the whole word.
    public static let schemaFiles: [String: String] = [
        "occurrent": "occurrent.schema.json",
        "causal_relation_object": "causal_relation_object.schema.json",
        "continuant": "continuant.schema.json",
        "realizable": "realizable.schema.json",
        "stratum": "stratum.schema.json",
        "bridge": "bridge.schema.json",
        "port": "port.schema.json",
        "conduit": "conduit.schema.json",
        "quality": "quality.schema.json",
        "token_individual": "individual.schema.json",
        "token_occurrence": "token.schema.json",
        "state_assertion": "state.schema.json",
        "token_causal_claim": "token_causal_claim.schema.json",
        "assertion": "assertion.schema.json",
        "enrichment": "enrichment.schema.json",
        "retraction": "retraction.schema.json",
        "succession": "succession.schema.json",
    ]

    /// The base URI under which every cross-file $ref names a sibling schema.
    private static let refBase = "https://causalontology.org/schema/"

    private let schemaDirectory: URL
    private var cache: [String: JsonValue] = [:]
    /// filename -> parsed schema, for cross-file $ref resolution.
    private var fileCache: [String: JsonValue] = [:]

    public init(schemaDirectory: URL) {
        self.schemaDirectory = schemaDirectory
    }

    /// The default spec/schema directory: the CAUSALONTOLOGY_SPEC environment
    /// variable (a spec/ directory) if set, else the CAUSALONTOLOGY_ROOT
    /// environment variable (the repository root) if set, else the location
    /// derived from this source file's position in the repository
    /// (bindings/swift/Sources/Causalontology/ -> five parents up -> root).
    public static func defaultSchemaDirectory() -> URL {
        let environment = ProcessInfo.processInfo.environment
        if let spec = environment["CAUSALONTOLOGY_SPEC"] {
            return URL(fileURLWithPath: spec).appendingPathComponent("schema")
        }
        if let root = environment["CAUSALONTOLOGY_ROOT"] {
            return URL(fileURLWithPath: root)
                .appendingPathComponent("spec")
                .appendingPathComponent("schema")
        }
        var url = URL(fileURLWithPath: #filePath)
        // SchemaValidator.swift -> Causalontology -> Sources -> swift
        // -> bindings -> repository root.
        for _ in 0..<5 {
            url.deleteLastPathComponent()
        }
        return url.appendingPathComponent("spec").appendingPathComponent("schema")
    }

    /// A shared validator over the default schema directory.
    public static let standard = SchemaValidator(schemaDirectory: defaultSchemaDirectory())

    /// Load (and cache) the root schema for a kind.
    public func loadSchema(forKind kind: String) throws -> JsonValue {
        if let cached = cache[kind] {
            return cached
        }
        guard let fileName = SchemaValidator.schemaFiles[kind] else {
            throw CausalontologyError("unknown kind: \(kind)")
        }
        let url = schemaDirectory.appendingPathComponent(fileName)
        let data: Data
        do {
            data = try Data(contentsOf: url)
        } catch {
            throw CausalontologyError("cannot read schema file: \(url.path)")
        }
        let schema = try JsonValue.parse(data)
        cache[kind] = schema
        return schema
    }

    /// (ok, reasons) - structural validity against the kind's JSON Schema.
    public func validate(
        _ obj: [String: JsonValue],
        kind: String? = nil
    ) throws -> (ok: Bool, reasons: [String]) {
        let resolvedKind: String
        if let kind = kind {
            resolvedKind = kind
        } else {
            resolvedKind = try inferKind(obj)
        }
        let root = try loadSchema(forKind: resolvedKind)
        var errors: [String] = []
        try check(.object(obj), schema: root, root: root, path: "$", errors: &errors)
        return (errors.isEmpty, errors)
    }

    // MARK: - The keyword interpreter

    /// Load (and cache) a sibling schema file by name, for cross-file $refs.
    private func loadFile(_ fileName: String) throws -> JsonValue {
        if let cached = fileCache[fileName] {
            return cached
        }
        let url = schemaDirectory.appendingPathComponent(fileName)
        let data: Data
        do {
            data = try Data(contentsOf: url)
        } catch {
            throw CausalontologyError("cannot read schema file: \(url.path)")
        }
        let schema = try JsonValue.parse(data)
        fileCache[fileName] = schema
        return schema
    }

    /// Navigate a JSON pointer ("$defs/interval") within a document.
    private func navigate(_ document: JsonValue, pointer: Substring) throws -> JsonValue {
        var node = document
        for part in pointer.split(separator: "/") {
            guard let next = node[String(part)] else {
                throw CausalontologyError("unresolvable pointer: \(pointer)")
            }
            node = next
        }
        return node
    }

    /// Follow local ("#/$defs/...") and cross-file
    /// ("https://causalontology.org/schema/<file>.schema.json#/...") $ref
    /// chains, returning both the resolved subschema and its (possibly new)
    /// root document, exactly as the Python binding's _resolve does.
    private func resolveRef(
        _ schema: JsonValue,
        root: JsonValue
    ) throws -> (schema: JsonValue, root: JsonValue) {
        var current = schema
        var currentRoot = root
        while let ref = current["$ref"]?.stringValue {
            if ref.hasPrefix("#/") {
                current = try navigate(currentRoot, pointer: ref.dropFirst(2)[...])
            } else if ref.hasPrefix(SchemaValidator.refBase) {
                let rest = ref.dropFirst(SchemaValidator.refBase.count)
                let fileName: String
                let pointer: Substring
                if let hash = rest.range(of: "#/") {
                    fileName = String(rest[rest.startIndex..<hash.lowerBound])
                    pointer = rest[hash.upperBound...]
                } else {
                    fileName = String(rest)
                    pointer = ""
                }
                currentRoot = try loadFile(fileName)
                current = pointer.isEmpty
                    ? currentRoot
                    : try navigate(currentRoot, pointer: pointer)
            } else {
                throw CausalontologyError("unsupported $ref: \(ref)")
            }
        }
        return (current, currentRoot)
    }

    /// True when the value is an instance of the named JSON Schema type.
    private func typeMatches(_ value: JsonValue, _ typeName: String) -> Bool {
        switch typeName {
        case "object":
            return value.objectValue != nil
        case "array":
            return value.arrayValue != nil
        case "string":
            return value.stringValue != nil
        case "number":
            // Booleans are never numbers here (JsonValue keeps them apart).
            return value.numberValue != nil
        case "integer":
            // Only integer literals; a floating literal is not an integer,
            // exactly as Python's isinstance(value, int) rejects a float.
            return value.intValue != nil
        case "boolean":
            return value.boolValue != nil
        case "null":
            return value.isNull
        default:
            return false
        }
    }

    /// A short display form for error messages.
    private func describe(_ value: JsonValue) -> String {
        switch value {
        case let .string(text):
            return "'\(text)'"
        default:
            return (try? jcsString(value)) ?? "<value>"
        }
    }

    /// True when the pattern is found anywhere in the string (re.search
    /// semantics; every Causalontology pattern is itself ^...$-anchored).
    private func patternFound(_ pattern: String, in text: String) throws -> Bool {
        let regex: NSRegularExpression
        do {
            regex = try NSRegularExpression(pattern: pattern)
        } catch {
            throw CausalontologyError("invalid schema pattern: \(pattern)")
        }
        let range = NSRange(text.startIndex..<text.endIndex, in: text)
        return regex.firstMatch(in: text, options: [], range: range) != nil
    }

    /// The recursive keyword check; appends messages to errors.
    private func check(
        _ value: JsonValue,
        schema rawSchema: JsonValue,
        root: JsonValue,
        path: String,
        errors: inout [String]
    ) throws {
        let (schema, root) = try resolveRef(rawSchema, root: root)

        if let branches = schema["oneOf"]?.arrayValue {
            var passing = 0
            for branch in branches {
                var branchErrors: [String] = []
                try check(value, schema: branch, root: root, path: path, errors: &branchErrors)
                if branchErrors.isEmpty {
                    passing += 1
                }
            }
            if passing != 1 {
                errors.append("\(path): matches \(passing) of the oneOf branches (need exactly 1)")
            }
            return
        }

        if let typeName = schema["type"]?.stringValue {
            if !typeMatches(value, typeName) {
                errors.append("\(path): expected \(typeName)")
                return
            }
        }

        if let constValue = schema["const"], value != constValue {
            errors.append("\(path): must equal \(describe(constValue))")
        }
        if let allowed = schema["enum"]?.arrayValue, !allowed.contains(value) {
            errors.append("\(path): \(describe(value)) not in enumeration")
        }
        if let pattern = schema["pattern"]?.stringValue, let text = value.stringValue {
            if try !patternFound(pattern, in: text) {
                errors.append("\(path): '\(text)' does not match \(pattern)")
            }
        }
        if let minLength = schema["minLength"]?.numberValue, let text = value.stringValue {
            if Double(text.unicodeScalars.count) < minLength {
                errors.append("\(path): shorter than minLength")
            }
        }
        if let minimum = schema["minimum"]?.numberValue, let number = value.numberValue {
            if number < minimum {
                errors.append("\(path): below minimum \(describe(schema["minimum"]!))")
            }
        }
        if let maximum = schema["maximum"]?.numberValue, let number = value.numberValue {
            if number > maximum {
                errors.append("\(path): above maximum \(describe(schema["maximum"]!))")
            }
        }

        if let items = value.arrayValue {
            if let minItems = schema["minItems"]?.numberValue {
                if Double(items.count) < minItems {
                    errors.append("\(path): fewer than \(Int(minItems)) items")
                }
            }
            if let itemSchema = schema["items"] {
                for (index, item) in items.enumerated() {
                    try check(item, schema: itemSchema, root: root,
                              path: "\(path)[\(index)]", errors: &errors)
                }
            }
        }

        if let members = value.objectValue {
            let properties = schema["properties"]?.objectValue ?? [:]
            if let required = schema["required"]?.arrayValue {
                for requirement in required {
                    if let name = requirement.stringValue, members[name] == nil {
                        errors.append("\(path): required property '\(name)' missing")
                    }
                }
            }
            if schema["additionalProperties"] == JsonValue.bool(false) {
                for key in members.keys.sorted() where properties[key] == nil {
                    errors.append("\(path): additional property '\(key)'")
                }
            }
            for key in properties.keys.sorted() {
                if let member = members[key] {
                    try check(member, schema: properties[key]!, root: root,
                              path: "\(path).\(key)", errors: &errors)
                }
            }
        }
    }
}

/// (ok, reasons) against the shared default validator - the module-level
/// convenience mirroring the Python binding's validate_schema.
public func validateSchema(
    _ obj: [String: JsonValue],
    kind: String? = nil
) throws -> (ok: Bool, reasons: [String]) {
    return try SchemaValidator.standard.validate(obj, kind: kind)
}

File: bindings/php/src/SchemaValidator.php
<?php

/* Schema validation against spec/schema/*.schema.json.
 *
 * A deliberately small interpreter for exactly the JSON Schema keywords the
 * eight Causalontology schemas use: type, const, enum, pattern, required,
 * properties, additionalProperties, items, minItems, minLength, minimum,
 * maximum, oneOf, and local $ref (#/$defs/...). "format" is treated as an
 * annotation, as the 2020-12 draft does by default.
 */

declare(strict_types=1);

namespace Causalontology;

final class SchemaValidator
{
    /** The schema file per kind, under spec/schema/. */
    public const SCHEMA_FILES = [
        'causal_relation_object'        => 'cro.schema.json',
        'occurrent'  => 'occurrent.schema.json',
        'continuant' => 'continuant.schema.json',
        'realizable' => 'realizable.schema.json',
        'assertion'  => 'assertion.schema.json',
        'enrichment' => 'enrichment.schema.json',
        'retraction' => 'retraction.schema.json',
        'succession' => 'succession.schema.json',
    ];

    /** @var array<string, array> loaded schemas, one per kind */
    private static array $cache = [];

    /** A static utility class, never an instance. */
    private function __construct()
    {
    }

    /** The spec/schema directory: env override, else the repository's. */
    private static function schemaDir(): string
    {
        $env = getenv('CAUSALONTOLOGY_SPEC');
        if (is_string($env) && $env !== '') {
            return $env . '/schema';
        }
        // src/SchemaValidator.php -> bindings/php/src, three levels below root.
        return dirname(__DIR__, 3) . '/spec/schema';
    }

    /** Load (and cache) the JSON Schema for a kind. */
    public static function loadSchema(string $kind): array
    {
        if (!isset(self::SCHEMA_FILES[$kind])) {
            throw new \InvalidArgumentException('unknown kind: ' . var_export($kind, true));
        }
        if (!isset(self::$cache[$kind])) {
            $file = self::schemaDir() . '/' . self::SCHEMA_FILES[$kind];
            $raw = @file_get_contents($file);
            if ($raw === false) {
                throw new \RuntimeException('cannot read schema file: ' . $file);
            }
            self::$cache[$kind] = json_decode($raw, true, 512, JSON_THROW_ON_ERROR);
        }
        return self::$cache[$kind];
    }

    /**
     * [ok, reasons] - structural validity against the kind's JSON Schema.
     *
     * @return array{0: bool, 1: list<string>}
     */
    public static function validateSchema(array $obj, ?string $kind = null): array
    {
        $kind ??= Canonical::inferKind($obj);
        $root = self::loadSchema($kind);
        $errors = [];
        self::check($obj, $root, $root, '$', $errors);
        return [$errors === [], $errors];
    }

    /** Follow local $ref chains (#/$defs/...) to the referenced schema. */
    private static function resolveRef(array $schema, array $root): array
    {
        while (array_key_exists('$ref', $schema)) {
            $ref = $schema['$ref'];
            if (!str_starts_with($ref, '#/')) {
                throw new \InvalidArgumentException('only local $ref supported: ' . $ref);
            }
            $node = $root;
            foreach (explode('/', substr($ref, 2)) as $part) {
                $node = $node[$part];
            }
            $schema = $node;
        }
        return $schema;
    }

    /** Does the value match a JSON Schema primitive type name? */
    private static function typeMatches(mixed $value, string $type): bool
    {
        return match ($type) {
            // An empty decoded array counts as a list, never a map - see the
            // representational note in Jcs.php (our data has no empty object).
            'object'  => Jcs::isMap($value),
            'array'   => Jcs::isList($value),
            'string'  => is_string($value),
            // PHP booleans are a distinct type, so no bool-as-number leak.
            'number'  => is_int($value) || is_float($value),
            'boolean' => is_bool($value),
            default   => throw new \InvalidArgumentException('unsupported schema type: ' . $type),
        };
    }

    /** Recursive keyword interpreter; appends messages to $errors. */
    private static function check(mixed $value, array $schema, array $root,
                                  string $path, array &$errors): void
    {
        $schema = self::resolveRef($schema, $root);

        if (array_key_exists('oneOf', $schema)) {
            $passing = 0;
            foreach ($schema['oneOf'] as $subSchema) {
                $subErrors = [];
                self::check($value, $subSchema, $root, $path, $subErrors);
                if ($subErrors === []) {
                    $passing++;
                }
            }
            if ($passing !== 1) {
                $errors[] = sprintf('%s: matches %d of the oneOf branches (need exactly 1)',
                                    $path, $passing);
            }
            return;
        }

        if (array_key_exists('type', $schema)) {
            if (!self::typeMatches($value, $schema['type'])) {
                $errors[] = $path . ': expected ' . $schema['type'];
                return;
            }
        }

        if (array_key_exists('const', $schema) && !Jcs::equal($value, $schema['const'])) {
            $errors[] = $path . ': must equal ' . json_encode($schema['const']);
        }
        if (array_key_exists('enum', $schema)) {
            $found = false;
            foreach ($schema['enum'] as $candidate) {
                if (Jcs::equal($value, $candidate)) {
                    $found = true;
                    break;
                }
            }
            if (!$found) {
                $errors[] = $path . ': ' . json_encode($value) . ' not in enumeration';
            }
        }
        if (array_key_exists('pattern', $schema) && is_string($value)) {
            // re.search semantics: an unanchored match; '~' never occurs in
            // the eight schemas' patterns, so it is a safe delimiter.
            if (preg_match('~' . $schema['pattern'] . '~', $value) !== 1) {
                $errors[] = $path . ': ' . json_encode($value)
                          . ' does not match ' . $schema['pattern'];
            }
        }
        if (array_key_exists('minLength', $schema) && is_string($value)) {
            // Byte length: equivalent to character length for the only
            // constraint the schemas carry (minLength 1).
            if (strlen($value) < $schema['minLength']) {
                $errors[] = $path . ': shorter than minLength';
            }
        }
        if (array_key_exists('minimum', $schema) && (is_int($value) || is_float($value))) {
            if ($value < $schema['minimum']) {
                $errors[] = $path . ': below minimum ' . $schema['minimum'];
            }
        }
        if (array_key_exists('maximum', $schema) && (is_int($value) || is_float($value))) {
            if ($value > $schema['maximum']) {
                $errors[] = $path . ': above maximum ' . $schema['maximum'];
            }
        }

        if (Jcs::isList($value)) {
            if (array_key_exists('minItems', $schema) && count($value) < $schema['minItems']) {
                $errors[] = sprintf('%s: fewer than %d items', $path, $schema['minItems']);
            }
            if (array_key_exists('items', $schema)) {
                foreach ($value as $index => $item) {
                    self::check($item, $schema['items'], $root,
                                $path . '[' . $index . ']', $errors);
                }
            }
        }

        if (Jcs::isMap($value)) {
            $properties = $schema['properties'] ?? [];
            foreach ($schema['required'] ?? [] as $required) {
                if (!array_key_exists($required, $value)) {
                    $errors[] = $path . ": required property '" . $required . "' missing";
                }
            }
            if (($schema['additionalProperties'] ?? null) === false) {
                foreach (array_keys($value) as $key) {
                    if (!array_key_exists((string) $key, $properties)) {
                        $errors[] = $path . ": additional property '" . $key . "'";
                    }
                }
            }
            foreach ($properties as $key => $subSchema) {
                if (array_key_exists($key, $value)) {
                    self::check($value[$key], $subSchema, $root,
                                $path . '.' . $key, $errors);
                }
            }
        }
    }
}
